### ISSUE TITLE
bool variables should allow `--verb:true` and `--verb=false` etc

### ISSUE THREAD

**Issue by timotheecour** [c-blake/cligen#16]
```
proc foobar(foo=1, bar=2.0, baz="hi", verb=false, paths: seq[string]): int =
  ##Some existing API call
  result = 1          # Of course, real code would have real logic here

when isMainModule: import cligen; dispatch(foobar)    # Whoa...Just one line??
```

I don't like toggles that can't be unset; eg it makes negating them or changing defaults awkward.
eg use case: if a default configuration uses `--verb`, there's no way to unset it.

the following would be backward compatible:

`--verb:FOO` => verb=parseBool(FOO)
`--verb=FOO` => verb=parseBool(FOO)

When `--verb` isn't followed by `:` or `=` we are still backward compatible:
for `--verb FOO ...`, make `FOO` part of next argument, eg `paths` in code below
for `--verb` (and no more arguments), set toggle as before: verb=true

```
proc foobar(foo=1, bar=2.0, baz="hi", verb=false, paths: seq[string]): int =
  ##Some existing API call
  result = 1          # Of course, real code would have real logic here

when isMainModule: import cligen; dispatch(foobar)    # Whoa...Just one line??
```

**Comment by c-blake** [comment]
The problem is not that you cannot "unset it"..It's a toggle.  So, you just repeat the ``--verb`` to unset it.  It does require that the user knows what the state is to be modified, though.  Sometimes wrapper scripts and aliases can obscure that.

I will have to look over the ``parseopt3.do_short`` and ``.do_long`` logic and see how easy it is to make a "NoVal" option key become more of an "OptionallyNoVal" in the way you suggest.  If you get to it before me (or have already done it), we can work on a PR.

**Comment by timotheecour** [comment]
actually i have another idea: make bool behave like all other types (ie not a toggle), but introduce another type that translates to a toggle:
 type toggle=distinct bool
its not good for generic code that bool behaves different from other types in many ways, eg setting it twice unsets, etc.

**Comment by timotheecour** [comment]
also note that usually, cmd line tools dont have this toggle behavior: setting twice doesnt unset. which tools have that behavior?

**Comment by c-blake** [comment]
I agree that "usually" some ``bool``-switched behavior uses a proc parameter that is false by default and activated by passing true. However, there are useful times when the default is true in an API call.  So, insisting the API call only default ``bool`` to ``false`` is a non-starter.  Instead, let's focus on the case when the default is ``true``.

The only useful meaning for presence/absence on a command-line can be to change it back to ``false``.  So, basically that _has_ to be supported or else cligen should not even allow changing such a parameter at all.  It doesn't need to be an endlessly flipping toggle, though.  Internally, I don't record the default or history of values bound to parameters.  So, it was simplest to just act as a toggle.

A cligen-defined ``toggle=distinct bool`` might work, but then the API author/maintainer needs to know about cligen which is tighter coupling than I would prefer.

I think it would be better to record some ``defaultsToTrue: set[string]`` (or maybe ``toggles``) and consult that.  The ``argParse`` template needs to receive this ``set[string]`` of long keys somehow, maybe as a new argument or maybe its existing ``val`` (since for that one type there currently is no value -- until we implement your above OptionalNoVal idea in parseopt3 anyway).  Either  way we need to adapt cligen to call ``argParse`` differently for ``bool``.  Or it might make sense for all ``argParse`` calls to receive two new arguments - default value as well as the current value - which would probably help with your "+=" idea.

Since ``argParse`` and ``argHelp`` are user-implemented for any user-defined types, I have been trying to keep them "easy".  Many or complex parameters can work against that.  I already think the first parameter to ``argHelp`` too complicated.  That said, a _(default, current, new)_ triple to ``argParse`` may not be so bad.  What do you think?

**Comment by c-blake** [comment]
Note also that, this sort of "change the ``argParse`` interface" approach impacts everyone using cligen with user-defined parsers or types and breaks their code since the template will get invoked with more parameters than their code expects.  I honestly don't know how many users cligen has (only how many github stars I get) or how impactful that would be. I am willing to make a breaking change like that for both ``bool`` and ``+/+=`` type features in the name of clean design, but we should really try to do it only once for cleanliness/featurefulness and so get it right.

**Comment by timotheecour** [comment]
> A cligen-defined toggle=distinct bool might work, but then the API author/maintainer needs to know about cligen which is tighter coupling than I would prefer.

indeed, I agree, forget about that `distinct bool` idea.

* I still don't like the idea of a toggle, I think we should get completely rid of toggle behavior. It'll be so easy to mis-use for end users in complex setups (eg when multiple default configs set default arguments for an app).

Also, leads to very unintuitive cmd lines, eg:
```
proc app(bool use_colors = true)= discard
./app # will use use_colors=true
./app --use_colors # will use use_colors=false, WTF?
```

* Here's my latest suggestion, I think it has all benefits we want:
make 
```
./app # use_colors=true (or false, depending on the default used for that param in the proc)
./app --use_colors # use_colors=true
./app --use_colors:false #use_colors=false (likewise with --use_colors=false)
./app --use_colors:true #use_colors=true (likewise with --use_colors=true)
./app --use_colors arg1 arg2 # use_colors=true; even when arg1 is `true` : arg1 is not parsed as a value for  `use_colors`
# with short form: no surprise there
./app -u # use_colors=true
./app -uv # use_colors=true, verify_ssl=true assuming `verify_ssl` is also a bool flag
./app -u:false -v
```

* it's intuitive
* it allows overriding a bool flag set to true
* no toggle behavior
* allows emulating cmd lines from other unix tools that have short form (eg: `ls -alh`)

**Comment by c-blake** [comment]
Re: the distinct bool, it is, of course, possible for CLI authors with strong opinions about bool to adjust it as in https://github.com/c-blake/cligen/blob/master/test/FancyRepeats.nim.

 I like most things about your idea, but not that one has to be so verbose to flip a default true value to false, but 'f' or '0' could be alternate spellings, I guess.  It's at least a 5-10% occurrence in my experience to have true defaults.

**Comment by timotheecour** [comment]
ok how about everything as i mentioned above, but also allow 0 or 1 as alias to true,false?

-u:0 -v=1 --baz=true --foo=0
other short aliases form could be y n t f but not huge fan of these, no strong opinion tho

**Comment by c-blake** [comment]
I did something that is mostly your proposal, but not quite.   You kind of ignored/ran past my plea to focus on the trickier case when the default ``bool`` is ``true``.  I don't like presence/absence on the command-line always meaning true/false no matter what the default is.  That basically feels like overriding/second guessing whoever defaulted the ``bool`` in the API call to ``true`` in the first place, but they have more information than ``cligen``.

So, the current behavior is just like your proposal if the default value for a ``bool`` is ``false``.  However, if the default value for a ``bool`` is ``true``, then using the implied option without an argument will mean ``false``.  This makes the most sense to me - ``bool`` flag present but unspecified value means flip the default, while a specified value means go with whatever is specified.

If the CLI user wants to pin a value of ``true`` or ``false`` then they can easily do so with ``:f`` or ``:t``.  There is no toggling back & forth.  If ``followSymlink=true`` is the default, then ``-f, -ff, -fff`` all make dispatch pass ``followSymlink=false`` (and similarly with long options).  That ``followSymlink`` example comes from the ``os.nim`` module.

If you really hate this behavior for your own programs, well, you can override it by just defining your own ``argParse`` and doing some ``include`` post-``cligen`` import, but pre-``dispatch``.  I don't know how many CLI authors will be wrapping routines with ``bool`` defaulting to ``true``.
